crates/sui-core/src/checkpoints/checkpoint_output.rs

```diff
@@ -106,3 +107,47 @@ impl CertifiedCheckpointOutput for LogCheckpointOutput {
         Ok(())
     }
 }
+
+pub struct SendCheckpointToStateSync {
+    sender: mpsc::Sender<CertifiedCheckpointSummary>,
+}
+
+pub struct ForwardToStateSyncTask {
+    receiver: mpsc::Receiver<CertifiedCheckpointSummary>,
+}
+
+impl ForwardToStateSyncTask {
+    pub fn start(mut self, handle: sui_network::state_sync::Handle) {
+        tokio::spawn(async move {
+            while let Some(checkpoint) = self.receiver.recv().await {
+                handle
+                    .send_checkpoint(VerifiedCheckpoint::new_unchecked(checkpoint))
+                    .await;
+            }
+        });
+    }
+}
+
+impl SendCheckpointToStateSync {
+    pub fn new() -> (Self, ForwardToStateSyncTask) {
+        let (sender, receiver) = mpsc::channel(128);
+        (Self { sender }, ForwardToStateSyncTask { receiver })
+    }
+}
+
+#[async_trait]
+impl CertifiedCheckpointOutput for SendCheckpointToStateSync {
+    async fn certified_checkpoint_created(
+        &self,
+        summary: &CertifiedCheckpointSummary,
+    ) -> SuiResult {
+        info!(
+            "Certified checkpoint with sequence {} and digest {}",
+            summary.summary.sequence_number,
+            Hex::encode(summary.summary.digest())
+        );
+        self.sender.send(summary.to_owned()).await.unwrap();
```

Comment:
I wonder if logging error here is better? Is there anything that guarantees that on node termination checkpoint service and all it's tasks always stops before state sync?

Reply:
I think the way the dependencies are this shouldn't be a problem? This channel shouldn't ever error on sending because the other side wont close until this sender is first dropped. But i'm happy to just log the error if it exists instead if you'd prefer.